### PR TITLE
Superscalar frontend

### PR DESCRIPTION
**Pull request by xThaid**
- the frontend is able to fetch many instructions in one cycle
- the fetch unit can also redirect itself, when for example it decodes a JAL instruction
- I added a simple branch prediction (branches jumping backwards predict as taken)

Benchmarks: https://github.com/xThaid/coreblocks/actions/runs/8599901154

TODO:
- [x] Get rid of `riscvmodel` dependency - code in `instr.py` for encoding instructions doesn't work yet (#631)
- [x] Add randomized tests of the fetch unit
- [x] More docstrings
- [x] Move `Pipe` data structure somewhere else. (#639)
- [x] Use `ArrayLayouts` (#640)
- [x] Add metrics

### CODE REVIEW

**Comment by tilk** [comment]
This FENCE test fail is weird. Is this a LSU bug, which somehow wasn't detected by all the other tests?

File: coreblocks/func_blocks/fu/lsu/dummyLsu.py [hunk 1]
```diff
@@ -228,17 +228,18 @@ def elaborate(self, platform):
         m.submodules.requester = requester = LSURequester(self.gen_params, self.bus)
 
         m.submodules.requests = requests = Forwarder(self.fu_layouts.issue)
-        m.submodules.results = results = self.forwarder = FIFO(self.lsu_layouts.accept, 2)
+        m.submodules.results_noop = results_noop = self.forwarder = FIFO(self.lsu_layouts.accept, 2)
```
Comment (piotro888) [inline coreblocks/func_blocks/fu/lsu/dummyLsu.py:231]:
```suggestion
        m.submodules.results_noop = results_noop = FIFO(self.lsu_layouts.accept, 2)
```
`self.forwarder = FIFO(` :) ; `self.forwarder` is also unused

Comment (piotro888) [reply coreblocks/func_blocks/fu/lsu/dummyLsu.py:231]: Not here, I missed the separate PR

**Review by lekcyjna123 — APPROVED** [review]
Approved. I left few comments but they can be solved under different PR.

File: coreblocks/frontend/fetch/fetch.py [hunk 2]
```diff
@@ -26,231 +42,468 @@ def __init__(self, gen_params: GenParams, icache: CacheInterface, cont: Method)
...
-        self.perf_rvc = HwCounter("frontend.ifu.rvc", "Number of decompressed RVC instructions")
+        self.write = Method(
+            i=[("valid_mask", self.width), ("slots", ArrayLayout(from_method_layout(self.elem_layout), self.width))]
+        )
+        self.read = Method(o=elem_layout)
+        self.clean = Method()
 
-        # For now assume that the fetch block is 4 bytes long (a machine word).
-        assert self.gen_params.fetch_block_bytes == 4
+        self.clean.add_conflict(self.write, Priority.LEFT)
```
[the remark this answers is not in the document]
Comment (xThaid) [reply coreblocks/frontend/fetch/fetch.py:405]: `add_conflict` also prohibits the methods from running simultaneously. I am using this semantics in the fetch unit to guarantee that whenever we have a stall signal, no instruction can leave the fetch unit (even in the same cycle).

File: coreblocks/frontend/fetch/fetch.py [hunk 3]
```diff
@@ -26,231 +42,468 @@ def __init__(self, gen_params: GenParams, icache: CacheInterface, cont: Method)
...
-        self.icache = icache
-        self.cont = cont
+    Serializes many elements one-by-one in order and dispatches it to the consumer.
+    The module accepts a new batch of elements only if the previous batch was fully
+    consumed.
 
-        self.resume = Method(i=self.gen_params.get(FetchLayouts).resume)
-        self.stall_exception = Method()
-        self.stall_exception.add_conflict(self.resume, Priority.LEFT)
+    It is a temporary workaround for a fetch buffer until the rest of the core becomes
```
[the remark this answers is not in the document]
Comment (xThaid) [reply coreblocks/frontend/fetch/fetch.py:391]: If we have another use case, I'll be happy to move it.

**Pull request merged**